### PR TITLE
t.remove: changed behaviour of flags to remove stds and connected maps

### PR DESCRIPTION
**Pull request by lucadelu**
Implemented behaviour suggested in https://trac.osgeo.org/grass/ticket/3362#comment:1

### CODE REVIEW

**Comment by veroandreo** [comment]
I do not manage to test properly. I tried pulling your branch and compiling and also getting the latest master, compiling and 
applying your patch, but in both cases, when I try to create a time series, I get the following:
```
t.create type=strds temporaltype=absolute output=precip_abs title="Example" descr="Example"
Creates a space time dataset.

Usage:
 t.create inputs=name[,name,...] maps=name[,name,...] [type=name]
   semantictype=string title=string description=string [--help]
   [--verbose] [--quiet] [--ui]

Parameters:
        inputs   Name of the input space time datasets
          maps   Type of the output space time dataset
          type   Type of the input map
                 options: raster,vector,raster_3d
                 default: raster
  semantictype   Semantic type of the space time dataset
                 options: min,max,sum,mean
                 default: mean
         title   Title of the new space time dataset
   description   Description of the new space time dataset

ERROR: t.create: Sorry, <temporaltype> is not a valid parameter
ERROR: t.create: Sorry, <output> is not a valid parameter
ERROR: Value <strds> out of range for parameter <type>
        Legal range: raster,vector,raster_3d
ERROR: Required parameter <inputs> not set:
        (Name of the input space time datasets)
ERROR: Required parameter <maps> not set:
        (Type of the output space time dataset)
```
Did you update master before the changes? Something seems to be really messed up with the options of t.create. It expects inputs and maps while this is not required by t.create at all. Meanwhile, the source code looks good, i.e., options are right. 

Might it be something in the parser? Because the options that it is asking for seem those for t.register, and those of t.register as well as other modules (t.rast.aggregate asks for temporaltype) are also messed up even without your patch (i.e., in the latest master):

```
t.register --h

Assigns timestamps and registers raster, vector and raster3d maps in a space time dataset.

Usage:
 t.register [-i] [nprocs=value] [map=name] [type=name] [file=name]
   [start=string] [end=string] [unit=string] [increment=string]
   [separator=character] [--overwrite] [--help] [--verbose] [--quiet]
   [--ui]

Flags:
  -i   Create an interval (start and end time) in case an increment and the start time are provided

Parameters:
     nprocs   Number of threads for parallel computing <<<< !!! 
              default: 1
        map   Name of the input map
       type   Type of the input space time dataset
              options: strds,stvds,str3ds
              default: strds
       file   Input file with map names, one per line
      start   Valid start date and time of the first map
        end   Valid end date and time of all map
       unit   Time stamp unit
              options: years,months,days,hours,minutes,seconds
  increment   Time increment, works only in conjunction with start option
  separator   Field separator character of the input file
              default: pipe
```

So, I cannot test so far, because I cannot create a time series... I'll open a bug report for the options mess

**Comment by lucadelu** [comment]
Probably we could change the behaviour of `-d` flag, right now it delete everything without `-f` maybe we should add it to remove the data

File: temporal/t.remove/t.remove.py [hunk 1]
```diff
@@ -117,16 +116,28 @@ def main():
             dataset_list.append(dataset_name)
 
     statement = ""
-
     # Create the pygrass Module object for g.remove
     remove = pyg.Module("g.remove", quiet=True, flags="f", run_=False)
 
+    if not force and not clean:
+        grass.message(_("The following data base element files would be deleted:"))
```
Comment (veroandreo) [inline temporal/t.remove/t.remove.py:123]:
```suggestion
        grass.message(_("The following data base element files will be deleted:"))
```

File: temporal/t.remove/t.remove.py [hunk 2]
```diff
@@ -117,16 +116,28 @@ def main():
...
-            grass.message(_("Removing registered maps and %s" % type))
+        if not force and not clean:
+            grass.message(
+                _("{stds}: {gid}".format(stds=sp.get_type().upper(), gid=sp.get_id()))
+            )
+        if recursive or clean:
+            if not force and not clean:
+                grass.message(
+                    _(
+                        "The following maps of {stds} {gid} would be deleted:".format(
```
Comment (veroandreo) [inline temporal/t.remove/t.remove.py:136]:
```suggestion
                        "The following maps of {stds} {gid} will be deleted:".format(
```

File: temporal/t.remove/t.remove.py [hunk 3]
```diff
@@ -136,46 +147,51 @@ def main():
...
-
-    # Execute the collected SQL statenents
-    dbif.execute_transaction(statement)
-
-    dbif.close()
+    if not force and not clean:
+        grass.message(
+            _(
+                "Nothing removed. You must use the force flag (-{flag}) to actually "
+                "remove them. Exiting.".format(flag="f")
```
Comment (veroandreo) [inline temporal/t.remove/t.remove.py:188]:
```suggestion
                "remove them.".format(flag="f")
```

**Review by veroandreo — COMMENTED** [review]
I'm testing within NC, using the following:
```
r.mapcalc expression="prec_1 = 100"
r.mapcalc expression="prec_2 = 200"
r.mapcalc expression="prec_3 = 300"
r.mapcalc expression="prec_4 = 400"
r.mapcalc expression="prec_5 = 500"
r.mapcalc expression="prec_6 = 600"

t.create type=strds temporaltype=absolute output=precip_abs title="Example" descr="Example"
t.register -i type=raster input=precip_abs maps=prec_1,prec_2,prec_3,prec_4,prec_5,prec_6 start="2001-01-01" increment="1 months"
```
Now, `t.remove -r inputs=precip_abs` rises the message of using `-f` to actually remove. Fine. When I use `t.remove -rf inputs=precip_abs` everything is removed (stds and maps) as the original behaviour and there's no message printed. I thought `-rf` would unregister maps and remove them, but keep the stds empty. However, stds is removed and maps are still in the mapset... This seems odd...

When I recreate the stds and use `t.remove -d inputs=precip_abs`, everything is removed (i.e., stds and maps from the mapset) but no message printed...

**Comment by veroandreo** [comment]
```
r.mapcalc expression="prec_1 = 100"
r.mapcalc expression="prec_2 = 200"
r.mapcalc expression="prec_3 = 300"
r.mapcalc expression="prec_4 = 400"
r.mapcalc expression="prec_5 = 500"
r.mapcalc expression="prec_6 = 600"

t.create type=strds temporaltype=absolute output=precip_abs title="Example" descr="Example"
t.register -i type=raster input=precip_abs maps=prec_1,prec_2,prec_3,prec_4,prec_5,prec_6 start="2001-01-01" increment="1 months"

t.remove -r inputs=precip_abs # remove only raster maps
t.remove -rf inputs=precip_abs # force removal of raster maps
t.remove -d inputs=precip_abs # remove stds and maps
```

> Probably we could change the behaviour of `-d` flag, right now it delete everything without `-f` maybe we should add it to remove the data

yes, and it should also warn the user that both stds and maps will be removed...

File: temporal/t.remove/t.remove.py (hunk 1 above)
Comment (lucadelu) [reply temporal/t.remove/t.remove.py]: but the will not removed, but only with force they will be removed

File: temporal/t.remove/t.remove.py (hunk 3 above)
Comment (lucadelu) [reply temporal/t.remove/t.remove.py:188]: The string is coming from g.remove, we should think to update also there

**Comment by lucadelu** [comment]
> I'm testing within NC, using the following:
> 
> ```
> r.mapcalc expression="prec_1 = 100"
> r.mapcalc expression="prec_2 = 200"
> r.mapcalc expression="prec_3 = 300"
> r.mapcalc expression="prec_4 = 400"
> r.mapcalc expression="prec_5 = 500"
> r.mapcalc expression="prec_6 = 600"
> 
> t.create type=strds temporaltype=absolute output=precip_abs title="Example" descr="Example"
> t.register -i type=raster input=precip_abs maps=prec_1,prec_2,prec_3,prec_4,prec_5,prec_6 start="2001-01-01" increment="1 months"
> ```
> 
> Now, `t.remove -r inputs=precip_abs` rises the message of using `-f` to actually remove. Fine. When I use `t.remove -rf inputs=precip_abs` everything is removed (stds and maps) as the original behaviour and there's no message printed. I thought `-rf` would unregister maps and remove them, but keep the stds empty. However, stds is removed and maps are still in the mapset... This seems odd...
> 

but in the description in [trac](https://trac.osgeo.org/grass/ticket/3362#comment:1) the suggested behaviour is the same of the implemented one.

-f remove only stds
-fr remove stds and unregister maps from temporal dataset
-fd remove stds, unregister and remove maps from temporal dataset

**Comment by wenzeslaus** [comment]
> - remove only stds
> - remove stds and unregister maps from temporal dataset
> - remove stds, unregister and remove maps from temporal dataset

These 3 levels also make sense to me in this order, although the descriptions are little hard to understand. Does the following describe the 3 behaviors?

- remove only stds
- remove stds and unregister maps
- remove stds, unregister maps, and remove them from mapset

**Comment by lucadelu** [comment]
> > * remove only stds
> > * remove stds and unregister maps from temporal dataset
> > * remove stds, unregister and remove maps from temporal dataset
> 
> These 3 levels also make sense to me in this order, although the descriptions are little hard to understand. Does the following describe the 3 behaviors?
> 
>     * remove only stds
> 
>     * remove stds and unregister maps
> 
>     * remove stds, unregister maps, and remove them from mapset

yes

File: temporal/t.remove/t.remove.py (hunk 1 above)
Comment (veroandreo) [reply temporal/t.remove/t.remove.py]: Yeah, it's just that would doesn't sound good to me. Besides, after the list of maps, there's the instruction saying, add the -f flag to actually remove, no? Indeed, I'd change the "would" also in g.remove... Not so important anyway

File: temporal/t.remove/t.remove.py (hunk 3 above)
Comment (veroandreo) [reply temporal/t.remove/t.remove.py:193]: yes, indeed, i'd remove the "would" and "Exiting" from g.remove

**Comment by lucadelu** [comment]
@wenzeslaus I rebased my branch with origin/master but It added the files modified by others. Is there a way to fix it or I should create new branch and close this pull request

**Comment by wenzeslaus** [comment]
> I rebased my branch with origin/master but It added the files modified by others.

You didn't pushed your changes which was a good call. It gives you generally more options for fixing things.

> Is there a way to fix it...

In general, see [GitFixUm](https://sethrobertson.github.io/GitFixUm/) (no GitHub stuff, just Git).

In this case, I have limited knowledge of what actually happened, but what I have usually seen in these cases is that people put other peoples' commits on top of theirs instead of putting their commits on top of the other peoples' commits. I actually don't know how that happens nor the Git commands to recreate similar situation. The correct command should be `git fetch upstream && git rebase upstream/master` (see e.g., [SO](https://stackoverflow.com/a/7929499/574907)). Are there wrong instructions somewhere? Since we are now always squashing while merging PRs, doing merge instead of rebase is a fine choice too (I use both depending on the circumstances).

(option 0) The other possibility is that you are in the middle of rebase without any new commits and you are seeing all these unrelated files. This sometimes happens and I'm not sure what are the circumstances, but it somehow makes sense. You can try to go through that rebase to the end following the instructions Git is giving you, solving any conflicts if needed. At the end, you shouldn't see anything wrong. Commits should be in the right order, have correct hashes, and your commits will be on top with new hashes.

(options 1 and 2) Anyway, if you have correct old commits from master, then your commits, and then bunch of these new commits by others, you can delete all the new commits by _resetting to the last of your commits_. Or you can _checkout the last of your commits and create a new branch_ from there.

You should check that your master branch (local one and also the one in your for just to be sure) has the same commits as the one on OSGeo/grass. Note that it is not just the commit messages and content, but also the commit hashes must be the same.

> ...or I should create new branch...

(options 3 and 4) Creating a completely new branch from master branch is likely the more straightforward way, but possibly with more work. Then, getting the changes through some file copying and Meld or cherry-picks would be the way to go and especially manual file copying and Meld works even for somehow broken commits.

(options 5) However, the state of the remote branch (visible here in the PR) seems fine, so you can just _delete your local branch_ and then _checkout the remote branch_ as you would for a branch from someone else. This is the option I would most likely choose unless you have some local changes.

> ...and close this pull request

Also, even with a new branch you don't need to create new PR. You can just _force push_ your new branch to the original remote branch (push to a branch with different name).

File: temporal/t.remove/t.remove.html [hunk 4]
```diff
@@ -2,12 +2,13 @@ <h2>DESCRIPTION</h2>
...
+the relevant database entries. It can also unregister maps from temporal
+database using the recursive mode <b>-r</b> (recursive)
 <p>
 Optionally, also the raster, 3D raster and vector maps of the space time
-datasets can be removed from the current mapset using the <b>-r</b> (recursive)
-and <b>-f</b> (force) flags. This recursive removal only works if both flags
-are checked together (use <b>-rf</b>).
+datasets can be removed from the current mapset using the <b>-d</b> (delete)
+flag.
+All removals only works if <b>-f</b> (force) flag is used.
```
Comment (veroandreo) [inline temporal/t.remove/t.remove.html:11]:
```suggestion
All removals only work if <b>-f</b> (force) flag is used.
```

File: temporal/t.remove/t.remove.py [hunk 5]
```diff
@@ -53,12 +53,17 @@
...
 
 # %flag
 # % key: f
-# % description: Force recursive removing
+# % description: Force removal (required for actual deletion of files)
+# %end
+
+# %flag
+# % key: d
+# % description: Remove stds and maps from temporal framework and from mapset
```
Comment (veroandreo) [inline temporal/t.remove/t.remove.py:66]:
```suggestion
# % description: Remove stds, unregister maps from temporal database and delete them from mapset
```

File: temporal/t.remove/t.remove.py [hunk 6]
```diff
@@ -117,16 +116,33 @@ def main():
...
-            grass.message(_("Removing registered maps and %s" % type))
+        if not force:
+            grass.message(
+                _("{stds}: {gid}".format(stds=sp.get_type().upper(), gid=sp.get_id()))
+            )
+        if recursive or clean:
+            if not force:
+                if recursive:
+                    msg = (
+                        "The following maps of {stds} {gid} would be "
```
Comment (veroandreo) [inline temporal/t.remove/t.remove.py:136]:
```suggestion
                        "The following maps of {stds} {gid} will be "
```

File: temporal/t.remove/t.remove.py [hunk 7]
```diff
@@ -117,16 +116,33 @@ def main():
...
+                        "unregistered from temporal database:"
+                    )
+                elif clean:
+                    msg = (
+                        "The following maps of {stds} {gid} would be "
```
Comment (veroandreo) [inline temporal/t.remove/t.remove.py:141]:
```suggestion
                        "The following maps of {stds} {gid} will be "
```

File: temporal/t.remove/t.remove.py [hunk 8]
```diff
@@ -53,12 +53,17 @@
 
 # %flag
 # % key: r
-# % description: Remove all registered maps from the temporal and also from the spatial database
+# % description: Remove all registered maps from the temporal
```
Comment (veroandreo) [inline temporal/t.remove/t.remove.py:56]:
```suggestion
# % description: Remove stds and unregister maps from temporal database
```

**Review by veroandreo — CHANGES_REQUESTED** [review]
I tested all combinations:
```
t.remove precip_abs        # I get the warning to use -f, fine
t.remove -f precip_abs     # removes only stds, maps are still registered in the temporal database, fine
t.remove -r precip_abs     # I get the warning to use -f, fine
t.remove -rf precip_abs    # removes stds and unregister maps from temporal database, fine
t.remove -d precip_abs     # the advice to use -f also here to force removal is missing
t.remove -df precip_abs    # removes stds, unregister maps and remove maps
```
So, only the advice to use `-f` to force the removal is missing from the case of using `-d` alone.

**Comment by lucadelu** [comment]
> > I rebased my branch with origin/master but It added the files modified by others.
> 
> You didn't pushed your changes which was a good call. It gives you generally more options for fixing things.
> 
> > Is there a way to fix it...
> 
> In general, see [GitFixUm](https://sethrobertson.github.io/GitFixUm/) (no GitHub stuff, just Git).
> 
> In this case, I have limited knowledge of what actually happened, but what I have usually seen in these cases is that people put other peoples' commits on top of theirs instead of putting their commits on top of the other peoples' commits. I actually don't know how that happens nor the Git commands to recreate similar situation. The correct command should be `git fetch upstream && git rebase upstream/master` (see e.g., [SO](https://stackoverflow.com/a/7929499/574907)). Are there wrong instructions somewhere? Since we are now always squashing while merging PRs, doing merge instead of rebase is a fine choice too (I use both depending on the circumstances).
> 
> (option 0) The other possibility is that you are in the middle of rebase without any new commits and you are seeing all these unrelated files. This sometimes happens and I'm not sure what are the circumstances, but it somehow makes sense. You can try to go through that rebase to the end following the instructions Git is giving you, solving any conflicts if needed. At the end, you shouldn't see anything wrong. Commits should be in the right order, have correct hashes, and your commits will be on top with new hashes.
> 
> (options 1 and 2) Anyway, if you have correct old commits from master, then your commits, and then bunch of these new commits by others, you can delete all the new commits by _resetting to the last of your commits_. Or you can _checkout the last of your commits and create a new branch_ from there.
> 
> You should check that your master branch (local one and also the one in your for just to be sure) has the same commits as the one on OSGeo/grass. Note that it is not just the commit messages and content, but also the commit hashes must be the same.
> 
> > ...or I should create new branch...
> 
> (options 3 and 4) Creating a completely new branch from master branch is likely the more straightforward way, but possibly with more work. Then, getting the changes through some file copying and Meld or cherry-picks would be the way to go and especially manual file copying and Meld works even for somehow broken commits.
> 
> (options 5) However, the state of the remote branch (visible here in the PR) seems fine, so you can just _delete your local branch_ and then _checkout the remote branch_ as you would for a branch from someone else. This is the option I would most likely choose unless you have some local changes.
> 
> > ...and close this pull request
> 
> Also, even with a new branch you don't need to create new PR. You can just _force push_ your new branch to the original remote branch (push to a branch with different name).

Ok, I tried to follow different suggestion, in the last to create a new branch and push to this one but I didn't succeed. 
I'm complete lost, I will create new one because is faster at this moment

**Pull request closed without merging**